cypress-io/cypress
# misc: Where possible allow Firefox and Chrome to atomically allocate ports (#25498)

Firefox and Chrome let the system allocate a port when started with `--remote-debugging-port=0` and announce the port on stderr afterwards where it is now parsed from.
    
Do something similar for Firefox's Marionette port and parse it from a message that gets written to stdout.
    
This doesn't seem possible with Firefox's `devtools.debugger.remote-port` so at least allow it to be overridden via preferences so users can implement their own race free port allocation mechanism when / where required:
```ts
            setupNodeEvents: (on, config) => {
                on('before:browser:launch', (browser, opts) => {
                    opts.preferences['devtools.debugger.remote-port'] = 12345
                    return opts;
                })
            }
```

- Closes #25498

### User facing changelog
- Allow users to override Firefox's `devtools.debugger.remote-port` via a `before:browser:launch` hook in `setupNodeEvents`.
This allows to work around issues which may result from automatically assigned / random ports.

### Steps to test

To test custom 'devtools.debugger.remote-port' override use the snippet from the description and verify that the given port is used.

For the remote debugging port in Chrome / Firefox verify port 0 is requested from the command line (`--remote-debugging-port=0`), verify a random port is used and that the connection can still be established. For Firefox's marionette port a 0 should be written to the `marionette.port` preference resulting in a random port as well.

### PR Tasks
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
